contracts/utils/HashUpdater.sol

```diff
@@ -0,0 +1,103 @@
+    /// @return multisig Service multisig address.
+    /// @return configHash IPFS hashes pointing to the config metapayload.
+    /// @return threshold Agent instance signers threshold.
+    /// @return maxNumAgentInstances Total number of agent instances.
+    /// @return numAgentInstances Actual number of agent instances.
+    /// @return state Service state.
+    function mapServices(uint256 serviceId) external view returns (uint96 securityDeposit, address multisig,
+        bytes32 configHash, uint32 threshold, uint32 maxNumAgentInstances, uint32 numAgentInstances, ServiceState state);
+
+    /// @dev Gets the owner of a specified service Id.
+    /// @param serviceId Service Id.
+    /// @return serviceOwner Service owner address.
+    function ownerOf(uint256 serviceId) external view returns (address serviceOwner);
+}
+
+/// @dev Provided zero address.
+error ZeroAddress();
+
+/// @dev Caught reentrancy violation.
+error ReentrancyGuard();
+
+/// @dev Account is unauthorized.
+/// @param account Account address.
+error UnauthorizedAccount(address account);
+
+/// @title HashUpdater - Metadata Hash Updater contract to manage actual / additional hashes
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract HashUpdater {
+    event HashUpdated(uint256 indexed serviceId, bytes32 indexed hash);
+
+    // Olas mech version number
+    string public constant VERSION = "0.1.0";
+
+    // Service Registry address
+    address public immutable serviceRegistry;
+
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Mapping of serviceId => actual metadata hash
+    mapping(uint256 => bytes32) public mapServiceHashes;
+
+    /// @dev HashUpdater constructor.
+    /// @param _serviceRegistry Service Registry address.
+    constructor(address _serviceRegistry) {
+        // Check for zero address
+        if (_serviceRegistry == address(0)) {
+            revert ZeroAddress();
+        }
+
+        serviceRegistry = _serviceRegistry;
+    }
+
+    /// @dev Changes metadata hash.
+    /// @param serviceId Service id.
+    /// @param hash Updated metadata hash.
+    function changeHash(uint256 serviceId, bytes32 hash) external {
+        // Reentrancy guard
+        if (_locked == 2) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Get service owner
+        address serviceOwner = IServiceRegistry(serviceRegistry).ownerOf(serviceId);
+        // Check for service owner
+        if (msg.sender != serviceOwner) {
+            // Get service multisig
+            (, address multisig, , , , , ) = IServiceRegistry(serviceRegistry).mapServices(serviceId);
+
+            // Check for service multisig address
+            if (msg.sender != multisig) {
+                revert UnauthorizedAccount(msg.sender);
+            }
```

Comment:
Imo this should be state dependent so that the service owner can update when they are in control and otherwise the agent service multisig itself can update.

Reply:
I believe this is exactly the case. When the service is not staked  - the service owner is always in control, otherwise it's multisig, as the staking contract being a service owner can't call this contract.